src/types.ts

```diff
@@ -54,6 +54,27 @@ export type ToTuple<T> = T extends [Parser<infer Head>, ...infer Tail]
   ? [Head, ...ToTuple<Tail>]
   : []
 
+/**
+ * Given a an array or a tuple of `Parser<T>`s, recursively extracts inner `T`s into a tuple or array.
+ *
+ * @example
+ *
+ * ```ts
+ * type U = [Parser<string>, Parser<number>, Parser<boolean>]
+ * type R = ToTuple<U> // type R = [string, number, boolean]
+ *
+ * type A = Parser<Array<string>>
+ * type T = ToTupleOrArray<A> // type T = string[]
+ * ```
+ */
+export type ToTupleOrArray<T> = T extends Array<Parser<infer Inner>>
+  ? Inner extends unknown
+    ? T extends [Parser<infer Head>, ...infer Tail]
+      ? [Head, ...ToTuple<Tail>]
+      : Inner[]
+    : []
+  : []
+
```

Comment:
For now it's only used in `sequence`. We need to discuss if it's valuable for deep inference or I completely misunderstood the api and usage. If that's the case we should probably constraint some of the usage further, so people can have `compile-time` errors instead of runtime.